doAssert, assert now print full path of failing line on error

timotheecour opened the pull request:
/cc @Araq 

before PR:
* when assert fails, no line context was shown;
* when doAssert fails, project path was shown (can be ambiguous), via `file(line)`

after PR: 
* doAssert, assert now print full path of failing line on error: `file(line, col)` (same rationale as https://github.com/nim-lang/Nim/pull/8466)
* refactoring so lib/core/macros.nim can reuse code to print Lineinfo

Varriount commented:
How will this affect tests that compare output of failed asserts?

timotheecour commented:
passes on travis but looks like I found a new bug on windows, which makes appveyor fail https://github.com/nim-lang/Nim/issues/8557 : it prints:

WARNING: C:\projects\nim\FAKEFILE(0, 6) false first assertion from bar
instead of:
WARNING: FAKEFILE(0, 6) false first assertion from bar

timotheecour commented:
found a great solution, PTAL

Araq commented on lib/system.nim:
```diff
@@ -3756,6 +3756,16 @@ proc instantiationInfo*(index = -1, fullPaths = false): tuple[
+  else:
+    file & "(" & $line & ", " & $column & ")"
+
+proc `$`*(info: type(instantiationInfo(0))): string =
```
Sorry, but no. Not two more system.nim entries.

timotheecour replied:
done, removed 2nd one; 1st one needed to keep things DRY (used in 3 places)

timotheecour commented:
PTAL

timotheecour commented:
ping: tests have passed

dom96 commented on lib/system.nim:
```diff
@@ -3756,6 +3756,13 @@ proc instantiationInfo*(index = -1, fullPaths = false): tuple[
+proc lineInfoToString*(file: string, line, column: int): string =
+  when defined(lineInfoToStringSquashForTesting):
+    # useful for for unittests, see tfailedassert.nim
+    "FAKEFILE"
```
This is just... please, just no.

timotheecour replied:
removed hardcoded FAKEFILE magic, see updated PR

dom96 commented on lib/system.nim:
```diff
@@ -3756,6 +3756,13 @@ proc instantiationInfo*(index = -1, fullPaths = false): tuple[
   ##     testException(IndexError, tester(1))
   ##     # --> Test failure at example.nim:20 with 'tester(1)'
 
+proc lineInfoToString*(file: string, line, column: int): string =
```
There is no universe in which this procedure should be in the ``system`` module

timotheecour replied:
removed, see how I did it in updated PR; I managed to avoid code duplication (this is used in both system.nim and macros.nim) while at same time not exposing new procs in system.nim

dom96 commented on lib/system.nim:
```diff
@@ -3791,10 +3805,8 @@ template doAssert*(cond: bool, msg = "") =
   {.line: instantiationInfo(-1, true).}:
-    if not cond:
-      raiseAssert(astToStr(cond) & ' ' &
-                  instantiationInfo(-1, false).fileName & '(' &
```
Why don't you just change this to ``true`` and be done with it?

timotheecour replied:
I don't understand your comment; that would change semantics

timotheecour replied:
simplified this, see updated PR

timotheecour commented:
PTAL. all comments addressed, and:
* added missing tests (doAssert wasn't tested)
* added test to check that condition `a+a=2` is stringified instead of becoming `false` in assert msg errr
* removed pre-existing code duplication between doAssert and assert
* avoided hardcoding FAKEFILE, instead: now if a test adds "-d:assertFailMsgMockFile=FOOBAR", assert msg error will show the file name mocked as FOOBAR ; using mocks is standard testing practice in software testing. Note the mock now ONLY affects error messages in doAssert, assert.

timotheecour commented:
PTAL @dom96 @Araq 
* fixes a regression introduced in https://github.com/nim-lang/Nim/pull/8731 which made doAssert not print the path
* fixes an off by one error that was pre-existing prior to #8731 in how the column was displayed
* doAssert and assert now print full path
* much simpler implementation removing all code duplication bw assert and doAssert
* added robust tests to make sure we won't introduce such regressions
* better mocking: instead of a compile time flag, we can inject at call site the desired path ; see how I handle that in tests: I test for both code paths (with real path and with mocked path; the real path ensures we won't have regressions in the future; the fake ones, used subsequently, are much easier to test)

dom96 commented:
I asked you to remove the ``FAKEFILE`` thing and you just moved it around. Is there a reason you cannot test this without adding a ``fakeLoc`` parameter to ``assert``?

timotheecour commented:
@dom96 PTAL ; I removed FAKEFILE and worked around it for the sake of moving forward with this PR. There's now a solid regression test in place that should catch future regressions, while at same time providing helpful messages when it fails so it's easy to fix if needed.
